src/modules/twap/hooks/useFetchDiscreteOrders.ts

```diff
@@ -0,0 +1,81 @@
+import { useMemo } from 'react'
+
+import type { Order } from '@cowprotocol/contracts'
+import { OrderParameters, SupportedChainId } from '@cowprotocol/cow-sdk'
+
+import { useAsyncMemo } from 'use-async-memo'
+
+import { ComposableCoW } from 'abis/types'
+import { computeDiscreteOrderUid } from 'utils/orderUtils/computeDiscreteOrderUid'
+
+import { TradeableOrderWithSignature, useTwapOrdersTradeableMulticall } from './useTwapOrdersTradeableMulticall'
+
+import { TwapDiscreteOrderItem, TwapDiscreteOrders } from '../state/twapDiscreteOrdersAtom'
+import { TwapOrderInfo } from '../types'
+
+export function useFetchDiscreteOrders(
+  safeAddress: string,
+  chainId: SupportedChainId,
+  composableCowContract: ComposableCoW,
+  ordersInfo: TwapOrderInfo[]
+): TwapDiscreteOrders | null {
+  const ordersToVerifyParams = useMemo(() => {
+    return ordersInfo.map((info) => info.safeData.params)
+  }, [ordersInfo])
+
+  const ordersTradeableData = useTwapOrdersTradeableMulticall(safeAddress, composableCowContract, ordersToVerifyParams)
+
+  const items = useAsyncMemo(
+    () => {
+      if (ordersInfo.length !== ordersTradeableData.length) return null
+
+      return Promise.all(
+        ordersInfo.map(({ id }, index) => {
+          const data = ordersTradeableData[index]
+          return data ? getTwapDiscreteOrderItem(chainId, safeAddress, data, id) : Promise.resolve(null)
+        })
+      )
+    },
+    [chainId, safeAddress, ordersInfo, ordersTradeableData],
+    null
+  )
+
+  return useMemo(() => {
+    if (!items) return null
+
+    return ordersInfo.reduce((acc, { id }, index) => {
+      const item = items[index]
+
+      if (item) acc[id] = item
+
+      return acc
+    }, {} as TwapDiscreteOrders)
+  }, [ordersInfo, items])
+}
+
+async function getTwapDiscreteOrderItem(
+  chainId: SupportedChainId,
+  safeAddress: string,
+  data: TradeableOrderWithSignature,
+  twapOrderId: string
+): Promise<TwapDiscreteOrderItem | null> {
+  if (!data) return null
+
+  const { order: discreteOrder, signature } = data
+  const fixedOrder = {
+    sellAmount: discreteOrder.sellAmount.toString(),
+    buyAmount: discreteOrder.buyAmount.toString(),
+    feeAmount: discreteOrder.feeAmount.toString(),
+    sellToken: discreteOrder.sellToken,
+    buyToken: discreteOrder.buyToken,
+    receiver: discreteOrder.receiver,
+    validTo: discreteOrder.validTo,
+    appData: discreteOrder.appData,
+    kind: 'sell', // TODO: discuss it, smart-contract returns bytes here
```

Comment:
It don't need to be always sell. Probably they return the bytes of the UTF8 encoding. We might need to decode. Not sure why is like this TBH

I know TWAPs are always SELL, but I believe this piece of code should be moved to `composableCow` module, since it's not for TWAPs but for any order in this framework.